sotodlib/io/g3tsmurf_utils.py

```diff
@@ -0,0 +1,195 @@
+    
+    Yields
+    --------
+    AxisManagers with loaded sections of data
+    """
+    
+    session = archive.Session()
+    obs = session.query(Observations).filter(Observations.obs_id==obs_id).one()
+    db_files = session.query(Files).filter(Files.obs_id==obs_id).order_by(Files.start)
+    filenames = sorted( [f.name for f in db_files])
+    
+    ts = obs.tunesets[0] ## if this throws an error we have some fallbacks
+    obs_dets, obs_samps = len(ts.dets), obs.n_samples
+    
+    logger.debug(f"{obs_id} has (n_dets, n_samps): ({obs_dets}, {obs_samps})")
+    if ram_limit is not None:
+        pts_limit = int(ram_limit // 4)
+        n_samp_chunks = 1
+        n_samps = obs_samps
+        n_dets = pts_limit // n_samps   
+        while n_dets == 0:
+            n_samp_chunks += 1
+            n_samps = obs_samps//n_samp_chunks
+            n_dets= pts_limit // (n_samps)
+        n_det_chunks = int(np.ceil( obs_dets/n_dets ))
+
+    if n_det_chunks is not None and n_dets is not None:
+        logger.warning("Both n_det_chunks and n_dets specified, n_det_chunks overrides")
+    if n_samp_chunks is not None and n_samps is not None:
+        logger.warning("Both n_samp_chunks and n_samps specified, n_samp_chunks overrides")
+        
+    if n_det_chunks is not None:
+        n_dets = int(np.ceil(obs_dets/n_det_chunks))
+        det_chunks = [range(i*n_dets,min((i+1)*n_dets,obs_dets)) for i in range(n_det_chunks)]
+    if n_samp_chunks is not None:
+        n_samps = int(np.ceil(obs_samps/n_samp_chunks))
+        samp_chunks = [(i*n_samps,min((i+1)*n_samps,obs_samps)) for i in range(n_samp_chunks)]
+
+    if n_dets is not None:
+        n_det_chunks = int(np.ceil(obs_dets/n_dets))
+        det_chunks = [range(i*n_dets,min((i+1)*n_dets,obs_dets)) for i in range(n_det_chunks)]
+    if n_samps is not None:
+        n_samp_chunks = int(np.ceil(obs_samps/n_samps))
+        samp_chunks = [(i*n_samps,min((i+1)*n_samps,obs_samps)) for i in range(n_samp_chunks)]
+        
+    if det_chunks is None:
+        det_chunks = [range(0,obs_dets)]
+    if samp_chunks is None:
+        samp_chunks = [(0,obs_samps)]
+        
+    ## should we let folks overwrite this here?
+    if "archive" in load_file_args:
+        archive = load_file_args.pop("archive")
+        
+    logger.debug(f"Loading data with det_chunks: {det_chunks}.")
+    logger.debug(f"Loading data in samp_chunks: {samp_chunks}.")
+    
+    try:
+        for det_chunk in det_chunks:
+            for samp_chunk in samp_chunks:
+                if test:
+                    yield (det_chunk, samp_chunk)
+                else:
+                    yield load_file( 
+                        filenames, 
+                        channels=det_chunk, 
+                        samples=samp_chunk, 
+                        archive=archive,
+                        **load_file_args,
+                    )
+    except GeneratorExit:
+        pass
+    session.close()
```

Comment:
I think the session should be closed immediately after you're done using it instead of waiting til after the data's processed

Reply:
I'm pretty sure that you technically want to leave it open until you're done with any objects you got from the database. But that's still well before this point so I'll move it up.